### PR TITLE
Beam 765 TriggerStateMachine does not need to be separate from OnceTriggerStateMachine

### PR DESCRIPTION
**Pull request by evindj**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by evindj** [comment]
@kennknowles Please do check and let me know if I need to change something.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterAllStateMachine.java [hunk 1]
```diff
@@ -78,32 +77,29 @@ public void onMerge(OnMergeContext c) throws Exception {
    */
   @Override
   public boolean shouldFire(TriggerContext context) throws Exception {
-    for (ExecutableTriggerStateMachine subtrigger : context.trigger().subTriggers()) {
-      if (!context.forTrigger(subtrigger).trigger().isFinished()
-          && !subtrigger.invokeShouldFire(context)) {
+    for (ExecutableTriggerStateMachine subtTrigger : context.trigger().subTriggers()) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterAllStateMachine.java]: `subtTrigger` looks like a typo

File: runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterAllStateMachine.java [hunk 2]
```diff
@@ -78,32 +77,29 @@ public void onMerge(OnMergeContext c) throws Exception {
...
-    for (ExecutableTriggerStateMachine subtrigger : context.trigger().subTriggers()) {
-      subtrigger.invokeOnFire(context);
+  public void onFire(TriggerContext context) throws Exception {
+    for (ExecutableTriggerStateMachine subTrigger : context.trigger().subTriggers()) {
+      subTrigger.invokeOnFire(context);
     }
+    context.trigger().setFinished(true);
   }
 
+
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/AfterAllStateMachine.java]: extra blank line

**Comment by kennknowles** [comment]
(very nice, by the way - I have no major comments because it is good)

**Comment by evindj** [comment]
Thank you so much for your feedback, You asked about removing OnceTriggerStateMachine, is there a specific reason I mean I can surely do that but I am hoping to understand more about the logic behind. Are we getting read of the OnceTrigger Semantic or is it already very well captured by the TriggerStateMachine implementation?

**Comment by kennknowles** [comment]
You have asked some very good questions about triggers.

First let me answer the question for this PR: The `TriggerStateMachine` is just a helper for a runner to implement triggers. It is not part of the definition of a trigger for a user. So `OnceTrigger` and `Trigger` exist in the SDK when you are building a pipeline, and the SDK handles validation there. When that pipeline arrives for execution, you don't have to keep track of it.

Separate from that, I do believe that we should eliminate the `OnceTrigger` conceptually, since it is a risk of data loss. But that work is somewhat more complex now because of backwards compatibility concerns. It is not part of this JIRA issue.

**Comment by evindj** [comment]
Thank you for your answers, it makes more sense to me now will remove  ```OnceTriggerStateMachine ``` implementation.

**Comment by kennknowles** [comment]
LGTM. Nice work jumping in to this code. I'll squash and merge.

**Pull request closed without merging**